File: 2021/python/Day-05/Day_05.py
import time

with open("input.txt","r") as file:
    input_list = []

    for line in (line.strip() for line in file):
        start, stop = (x for x in line.split(" -> "))
        start = tuple(int(x) for x in start.split(","))
        stop = tuple(int(x) for x in stop.split(","))
        input_list.append([start,stop])


def solution(input_lines, ignore_diagonal):
    drawn_lines = {}
    for line in input_lines:
        start, stop = line

        dir_x = -1 if start[0] > stop[0] else 1 if start[0]<stop[0] else 0
        dir_y = -1 if start[1] > stop[1] else 1 if start[1]<stop[1] else 0

        if ignore_diagonal and dir_x and dir_y:
            continue

        distance = abs(start[0] - stop[0]) or abs(start[1] - stop[1])

        for i in range(distance + 1):
            x = start[0] + dir_x * i
            y = start[1] + dir_y * i
            drawn_lines[(x,y)] = drawn_lines.get((x,y), 0) + 1

    return sum([1 for key in drawn_lines if drawn_lines[key] > 1])


solution01_start_time = time.time()
solution01 = solution(input_list, True)
solution01_total_time = time.time() - solution01_start_time
print(f"Solution 01: {solution01}, in {solution01_total_time}")

solution02_start_time = time.time()
solution02 = solution(input_list, False)
solution02_total_time = time.time() - solution02_start_time
print(f"Solution 02: {solution02}, in {solution02_total_time}")

File: 2021/js/Day-05/Day-05.js
const fs = require("fs");

inputData = fs.readFileSync("input.txt", "utf8");

inputData = inputData.split(/\r?\n/).map((v) => v.trim());

inputData = inputData.map((v) =>
  v
    .split(" -> ")
    .map((v) => v.split(","))
    .map((v) => v.map((v) => Number(v)))
);

console.log(inputData);

const solution = (lines, ignoreDiagonals = true) => {
  drawnLines = [];

  for (line of lines) {
    const [x1, y1, x2, y2] = line.flat();

    let dirX = x1 < x2 ? 1 : x1 > x2 ? -1 : 0;
    let dirY = y1 < y2 ? 1 : y1 > y2 ? -1 : 0;

    if (dirX && dirY && ignoreDiagonals) {
      continue;
    }

    //wonder if this will work:
    let distance = Math.abs(x1 - x2) || Math.abs(y1 - y2);

    for (let i = 0; i < distance + 1; i++) {
      x = x1 + dirX * i;
      y = y1 + dirY * i;

      drawnLines[[x, y]] = (drawnLines[[x, y]] || 0) + 1;
    }
  }
  // console.log(drawnLines);

  let counter = 0;

  console.log(counter);

  for (k in drawnLines) {
    if (drawnLines[k] > 1) counter++;
  }

  return counter;
};

console.log(`Solution 01: ${solution(inputData)}`);
console.log(`Solution 02: ${solution(inputData, false)}`);
